### ISSUE TITLE
🐞 [BUG] Error restoring a SMS backup copy

### ISSUE THREAD

**Issue by Ilya-Ezhik** [octoshrimpy/quik#70]
When trying to restore an SMS from a backup, an error occurs when reading the copy file.

1. Go to "Menu --> Backup"
2. Click on "Restore"
3. Select the copy file
4. Click "OK"

The dialog box will contain a description of the error "it is impossible to read the backup file"

 - Device: Samsung Galaxy A22 5G (as well as all other devices)
 - OS: Android 13, One UI Core 5.1 (and other Android OS
 - App Version 4.0.7 (latest)

![Screenshot_20240307_022132](https://github.com/octoshrimpy/quik/assets/130345901/11ce30c4-dc99-4d5c-9b3b-af8e3c6fffdc)

**Comment by octoshrimpy** [comment]
Looking now.

**Comment by Ilya-Ezhik** [comment]
> Looking now.

Thanks! There is also a problem with group SMS sending, but for this probably need to open a new topic?

**Comment by octoshrimpy** [comment]
> re is also a problem with group SMS sending, but for this probably need to open a new topic?

yes please.

**Comment by octoshrimpy** [comment]
looks like this may be because quik never asked for access to files, only media. it never writes a backup. I'm assuming it's the same problem when importing a backup. did this work with QKSMS?

**Comment by Ilya-Ezhik** [comment]
> it never writes a backup.

When creating a backup, it creates a JSON file in the selected folder on the phone with all SMS (headers from the phone number and SMS text)! And then he can't read it (described above)

**Comment by Ilya-Ezhik** [comment]
> Idid this work with QKSMS?

Yes, in that application, a backup is created and successfully restored! Without problems!

**Comment by octoshrimpy** [comment]
> > it never writes a backup.
> 
> When creating a backup, it creates a JSON file in the selected folder on the phone with all SMS (headers from the phone number and SMS text)! And then he can't read it (described above)

and you can open the json with another file viewer/editor without problem? I wonder if I have remnants of some debug build from testing directly from android studio that is interfering. mine isn't storing the file at all, which is even worse.

**Comment by Ilya-Ezhik** [comment]
> and you can open the json with another file viewer/editor without problem?

Yes! I open it on phone using the MT Manager app without any problems! :)

![Screenshot_20240308_002309_MT Manager](https://github.com/octoshrimpy/quik/assets/130345901/8ac95a48-3ab7-4936-83a5-7fe77c0965a1)

**Comment by edent** [comment]
I'm having the same issue. Both phones are use V4.0.7.

I have manually copied across the JSON file. When I try to open it in the app I get the error "Couldn't read backup file".

I can *create* a backup into a specific folder. But if I try to read that newly created backup on the same phone, I get that error.

**Comment by Ilya-Ezhik** [comment]
So what do we do now?

**Comment by esan0** [comment]
Same issue.

**Comment by engelhro** [comment]
Any news on this issue?

I'm affected as well. Switched from one phone to a new one today, and now I can't import my backup and all its messages. Would be really nice if this could be fixed 👍.

Thanks a lot for a short update!

**Comment by SanskritFritz** [comment]
Just a heads-up, a workaround: QKSMS is able to restore the backups made with QUIK.

**Comment by engelhro** [comment]
> Just a heads-up, a workaround: QKSMS is able to restore the backups made with QUIK.

Yeah, I've seen that afterwards and it indeed helped to solve my issue! 😀

The problem existing in QUIK but not already in QKSMS should hopefully help the developer to more easily identify and fix the problem.

Would still be nice to see a working restore functionality also in the "revived" app which is QUIK.

**Comment by b-risk** [comment]
Hi, I'm having a similar issue too, I made a backup with this revived app on my phone and I transferred it to my computer, after wiping my phone (so no making a new backup) and installing GrapheneOS and restoring from the file in settings it gives an error where it says "couldn't read"

**Comment by antistress** [comment]
Same bug here with latest version from F-Droid (4.0.7) on my new Galaxy S10e with lineage-20.0-20240608-nightly
And same workaround : I had to use QKSMS to import the backup file and then QUIK to synchronise itself with QKSMS

**Comment by Saarsk** [comment]
**Here was my workaround:**
I used the app "SMS Import / Export" from F-Droid. Readme here: https://github.com/tmo1/sms-ie?tab=readme-ov-file#readme

All I did was tap "import" and it appeared in QUIK. I see there's an option in QUIK to re-synch with the native SMS database.

**Comment by MoralCode** [comment]
> ### Here was my workaround:
> I used the app _"SMS Import / Export"_ available from [F-Droid](https://f-droid.org/packages/com.github.tmo1.sms_ie/).
> 
> You might want to check the [Readme](https://github.com/tmo1/sms-ie?tab=readme-ov-file#readme). There were no issues for me, but if you have group messages etc. there might be issues.
> 
> All I did was tap "import" and it appeared in QUIK. I see there's an option in QUIK to re-synch with the native SMS database.


This seemed to work for me, and also seemed to import MMS, which I suspect the built in importer doesn't do.

**Comment by rafnov** [comment]
Same on me.
Successfuly created backup .json (quite small - 1,17 MB) on Realme 7 Pro, uploaded to Google drive, then downloaded od Realme GT Neo 2 and tried to restore: unable to read a file. 😟
Any chances for soon fix? It's been some time since Quik was updated...

**Comment by rafnov** [comment]
Ok, it worked perfectly for me too, but that doesn't fix anything in Quik.

**Comment by Renarde-dev** [comment]
I could wait for a release and SMS-IE didn't work for me so I wrote a quick script to convert backups to Fossify Messages and used it to import my messages :
If you want to use my script it's available on my forgejo instance : https://forgejo.la-taniere-solidaire.gay/Renarde/SMS_Import_script

**Comment by dex2000** [comment]
FYI: it was broken in 4.06 too

**Comment by SzellenDienw** [comment]
Hello, it seems to me that QUIK can import .XML files.

**Comment by dex2000** [comment]
it should import his own backup-files...

**Comment by octoshrimpy** [comment]
I believe the issue is with permissions, as android permissions from when QKSMS was last updated vs now are very different. I'm trying to wrangle those permissions, as i'm having problems with camera as well.

**Comment by dex2000** [comment]
i dont think that changes of android permissions is the problem.

(Last qkms-update-> January> 23,
last update my android->Nov 20 ...
 qksms works as expected quik not)

**Comment by octoshrimpy** [comment]
you also have to take into consideration the version of android QKSMS was compiled against, vs quik. quik is compiled against a newer version of the SDK, one where the permissions were deprecated.

**Comment by MoralCode** [comment]
> i have compared now the permissions and see that qksms have additional

Where/how did you compare them? if you just looked at the permissions settings menus for each app in the UI of your device, theres a good chance that the underlying technical identifiers for each permission may have changed without being visible on that menu.

For example, i think newer versions of android allow more granular file access permissions, where, if an app needs access to photos, you can select groups, or even individual photos to allow access to, rather than it being "all or nothing" with your entire camera roll.

**Comment by dex2000** [comment]
> > i have compared now the permissions and see that qksms have additional
> 
> Where/how did you compare them? if you just looked at the permissions settings menus for each app in the UI of your device, 

Yep.

> theres a good chance that the underlying technical identifiers for each permission may have changed without being visible on that menu.

I'm a user not an app developer. 
I dont understand that two nearly identical Apps on same (!) one device are different working (or not working like here).

**Comment by MoralCode** [comment]
> I'm a user not an app developer. I dont understand that two nearly identical Apps on same (!) one device are different working (or not working like here).

It's okay If you aren't a dev! As someone who has played both roles, I figured it was worth asking clarifying, both to help octoshrimpy understand your point, and also to offer an explanation for how your observation and octoshrimpy's assertion that the permissions had changed could both be true at the same time.

**Comment by octoshrimpy** [comment]
@dex2000: @MoralCode is 100% correct. Any help is good help (and bad help is just Quality Control!) If you'd like to learn, here's a great place to start on the topic: 

[android SDK docs](https://developer.android.com/guide/topics/permissions/overview)
[quik's permissions](https://github.com/octoshrimpy/quik/blob/master/presentation/src/main/AndroidManifest.xml)

**Comment by ivan-lt** [comment]
I have the same issue, that is:
- install QUIK afresh, run it, agree to use it as a default messaging app;
- go to Backup and restore;
- set Backup location (I set it to Download/temp as I heard Android might be more permissive in the Download folder, don't know whether it's true);  at this time the OS asks "Allow QUIK to access files in temp?", to which I agree;
- tap Backup now and get a message "Backing up messages / Finished!";  the file is created and contains the following:
```
{
 "messageCount": 0,
 "messages": []
}
```
- tap Restore, get file selection screen, select the newly created file, and...
- get the Error / Couldn't read backup file.

I'm using QUIK 4.0.10-beta (Obtainium says it's 4.0.11), Pixel 7a, GrapheneOS.  It didn't ask for any additional permissions.  Didn't test QKSMS as the issue was not pressing but having a working import/export would definitely be nice.  Thanks for your work on this app!

**Comment by MoralCode** [comment]
> It didn't ask for any additional permissions.

what if you go into the app's info screen and give it as many filesystem related permissions as youre able to allow - does that change the behavior of whats included in the backup file?

**Comment by ivan-lt** [comment]
> what if you go into the app's info screen and give it as many filesystem related permissions as youre able to allow - 

There are "Music and audio" and "Photos and videos", and they are allowed.  In fact, it says "No permissions denied".

> does that change the behavior of whats included in the backup file?

Actually there's nothing wrong with the file's contents because it was a fresh install with no messages.  In fact, I've sent an SMS, made another backup and the message is duly reflected in the new file.  The "Couldn't read backup file" error persists, however.  (On both files.)

**Comment by dex2000** [comment]
@MoralCode i dont understand your answers here. 
Is it working at you? On which config?

Maybe it could be help the lot of people where its broken.

@ivan-lt all what i understand here (with my small english-knowledge): you have to learn programming languages and fix it yourself.

**Comment by dex2000** [comment]
I'm a noob and you know it, but:
https://github.com/octoshrimpy/quik/blob/4cbf3bdc4d1063c3cd2dd82626785bf8f279ffe8/domain/src/main/java/com/moez/QKSMS/repository/BackupRepository.kt#L53
looks very really logical, but in qksms are the last lines otherwise (stop comes first) any ideas?

**Comment by Inhishonor** [comment]
I was experimenting with this, and for some reason, when QUIK is not set as the default messaging app, it is able to restore the backup. This seems to confirm that permissions are the issue, though, I am not sure why being the default app is causing the problem.

**Comment by ivan-lt** [comment]
> I was experimenting with this, and for some reason, when QUIK is not set as the default messaging app, it is able to restore the backup.

That would be an awesome workaround, however, I can't reproduce it here -- even when QUIK is not set as the default app, it can't read backup files, neither old nor new.  Rebooting doesn't help.

**Comment by Inhishonor** [comment]
>  That would be an awesome workaround, however, I can't reproduce it here -- even when QUIK is not set as the default app, it can't read backup files, neither old nor new. Rebooting doesn't help.
What Android version are you using? In my reading regarding this, I found that Google recently changed the Android file permissions, which is what's causing this problem, I think.

**Comment by ivan-lt** [comment]
> What Android version are you using? 

Android 15 (GrapheneOS)

**Comment by Inhishonor** [comment]
I see, are you able to build this locally to test? After doing some research the issue may be with the [proguard rules](https://github.com/Inhishonor/quik/blob/master/presentation/proguard-rules.pro), they are the same as QKSMS but the code for Backup and Restore has changed quite a bit since then, and those rules are over 6 years old.

**Comment by oowais** [comment]
@Inhishonor you made me happy for a second, but it didn't work, I have the same problem as @ivan-lt mentioned. I am also using Graphene OS, Android 15.
Here are some logs that were collected during my process:
 - Install QUIK (4.0.7) from Driodify
 - Open app, and disallow it to be the default messaging app
 - Open backup and store settings menu
 - Set backup folder location
 - Select Restore option and select file
 - "Couldn't read backup file"

[QUIK log 7cc318b070d6.txt](https://github.com/user-attachments/files/18934364/QUIK.log.7cc318b070d6.txt)

**Comment by Inhishonor** [comment]
> @Inhishonor you made me happy for a second, but it didn't work, I have the same problem as @ivan-lt mentioned. I am also using Graphene OS, Android 15.
> Here are some logs that were collected during my process:
>  - Install QUIK (4.0.7) from Driodify
>  - Open app, and disallow it to be the default messaging app
>  - Open backup and store settings menu
>  - Set backup folder location
>  - Select Restore option and select file
>  - "Couldn't read backup file"
> 
> 
Sorry, I realized that when that happened, it was only because I had been running a locally built app. It was not due to the fact that I had revoked and then re-enabled default messaging. 
Also, thank you for the log. I will take a look at that and see if I can figure it out.

**Comment by Inhishonor** [comment]
I have narrowed it down to proguard as the culprit, with it disabled even on a release build, I was able to backup and restore.

**Comment by Inhishonor** [comment]
I submitted pull request #311 that modifies the proguard rules to allow the restore functionality to access the files. I confirmed this works across two devices, each running different android versions.
